### ISSUE TITLE
TLS handshake failure : getSelectedCipherSuite

### ISSUE THREAD

**Issue by SubhraB** [bcgit/bc-java#793]
I set up a spring boot application in bcfips mode.
```
		Security.insertProviderAt(new BouncyCastleFipsProvider(), 1);
		Security.insertProviderAt(new BouncyCastleJsseProvider(), 2);
		Security.removeProvider("SunRsaSign");
		Security.removeProvider("SunJSSE");
		Security.setProperty("ssl.KeyManagerFactory.algorithm", "PKIX");

```
This is how i generate server side keystore.
```
	public static byte[] pemToJks(String pemCert, String pemKey, String[] pemIntermediates, String jksPassword) {
		Security.insertProviderAt(new BouncyCastleFipsProvider(), 1);
		Security.removeProvider("SunRsaSign");

		try {
			try (PEMParser certParser = new PEMParser(new StringReader(pemCert));
					PEMParser keyParser = new PEMParser(new StringReader(pemKey));) {

				List<X509Certificate> x509Chain = new ArrayList<>();
				CertificateFactory certificateFactory = CertificateFactory.getInstance("X509");

				X509CertificateHolder holder = (X509CertificateHolder)certParser.readObject();
				x509Chain.add((X509Certificate) certificateFactory
						.generateCertificate(new ByteArrayInputStream(holder.getEncoded())));

				for (String pemIntermediate : pemIntermediates) {
					try (PEMParser intermediateParser = new PEMParser(new StringReader(pemIntermediate))) {
						holder = (X509CertificateHolder) intermediateParser.readObject();
						x509Chain.add((X509Certificate) certificateFactory
								.generateCertificate(new ByteArrayInputStream(holder.getEncoded())));
					}
				}

				PEMKeyPair pemKeyPair = (PEMKeyPair)keyParser.readObject();
				PrivateKey privateKey = new JcaPEMKeyConverter().getPrivateKey(pemKeyPair.getPrivateKeyInfo());

				KeyStore keystore = KeyStore.getInstance("BCFKS");
				keystore.load(null, null);
				String commonName = new NameReader(x509Chain.get(0)).readSubject().getValue(StandardAttributeType.CommonName);
				keystore.setKeyEntry(commonName, privateKey, jksPassword.toCharArray(),
						x509Chain.toArray(new X509Certificate[0]));

				ByteArrayOutputStream bos = new ByteArrayOutputStream();

				keystore.store(bos, jksPassword.toCharArray());
				bos.close();

				return bos.toByteArray();
			}

		} catch (Exception e) {
			logger.error(e.getMessage(), e);
			throw new RuntimeException(e);
		}
	}

```
BC versions-
```
	compile group: 'org.bouncycastle', name: 'bc-fips', version: '1.0.2'
	compile group: 'org.bouncycastle', name: 'bctls-fips', version: '1.0.10'

```

When I try to access my http endoint, I get the below error.
```
Oct 06, 2020 10:53:03 AM org.bouncycastle.jsse.provider.ProvTlsServer notifyAlertRaised
INFO: Server raised fatal(2) handshake_failure(40) alert: Failed to process record
org.bouncycastle.tls.TlsFatalAlert: handshake_failure(40)
	at org.bouncycastle.tls.AbstractTlsServer.getSelectedCipherSuite(AbstractTlsServer.java:387)
	at org.bouncycastle.jsse.provider.ProvTlsServer.getSelectedCipherSuite(ProvTlsServer.java:348)
	at org.bouncycastle.tls.TlsServerProtocol.generateServerHelloMessage(TlsServerProtocol.java:191)
	at org.bouncycastle.tls.TlsServerProtocol.handleHandshakeMessage(TlsServerProtocol.java:528)
	at org.bouncycastle.tls.TlsProtocol.processHandshakeQueue(TlsProtocol.java:644)
	at org.bouncycastle.tls.TlsProtocol.processRecord(TlsProtocol.java:546)

```

However, when I use simple keytool command to generate the keystore (with provider options),  it works just fine.

Also, with bctls-fips version 1.0.10.1, it works fine both keystores on my mac, but fails with the same error on a k8s container.

**Comment by peterdettman** [comment]
1.0.10.1 was specifically a DSA-related fix. Are the credentials DSA? The getSelectedCipherSuite failure is consistent with a failure to find suitable credentials in the keystore.

It would be very strange if an identical setup worked on a mac, but not in k8s. I suggest reviewing the configuration, especially the provider setup and "ssl.KeyManagerFactory.algorithm" setting, enabled cipher suites configuration, and keystore configuration.

Java Logging API output might be useful too (see Appendix C of the (D)TLS user guide).

**Comment by SubhraB** [comment]
Hi @peterdettman. Thanks for responding.

I believe bc-fips and bctls-fips are just fine. With keytool generated BCFKS, everything works just fine. I am not blocked anymore.

The issue might be with the way I was creating the keystore using the library `bcpkix-fips-1.0.4.jar`, that doesn't work at all. I am replacing that with simple keytool to create keystore. Here was the code I used.

Step 1: Generate key and csr
```
	public static Pair<String, String> generateKeyAndCsr(String commonName, List<String> subjectAlternativeNames,
			String email) {

		try {

			KeyPairGenerator keyPairGenerator = KeyPairGenerator.getInstance("RSA");
			keyPairGenerator.initialize(new RSAKeyGenParameterSpec(3072, RSAKeyGenParameterSpec.F4));
			KeyPair keypair = keyPairGenerator.generateKeyPair();

			PublicKey publicKey = keypair.getPublic();
			PrivateKey privateKey = keypair.getPrivate();

			StringWriter keyWriter = new StringWriter();
			JcaPEMWriter keyPemWriter = new JcaPEMWriter(keyWriter);
			keyPemWriter.writeObject(privateKey);
			keyPemWriter.close();

			ContentSigner contentSigner = new JcaContentSignerBuilder("SHA256withRSA").build(privateKey);

			X500NameBuilder x500NameBuider = new X500NameBuilder(BCStyle.INSTANCE);

			x500NameBuider.addRDN(BCStyle.C, "US");
			x500NameBuider.addRDN(BCStyle.ST, "Washington");
			x500NameBuider.addRDN(BCStyle.L, "Bellevue");
			x500NameBuider.addRDN(BCStyle.O, "Concur");
			x500NameBuider.addRDN(BCStyle.E, email);
			x500NameBuider.addRDN(BCStyle.CN, commonName);

			X500Name subject = x500NameBuider.build();

			PKCS10CertificationRequestBuilder requestBuilder = new JcaPKCS10CertificationRequestBuilder(subject,
					publicKey);

			if (CollectionUtils.isNotEmpty(subjectAlternativeNames)) {
				List<GeneralName> namesList = subjectAlternativeNames.stream()
						.map(san -> new GeneralName(GeneralName.dNSName, san)).collect(Collectors.toList());
				GeneralNames generalNames = new GeneralNames(namesList.toArray(new GeneralName[] {}));
				ExtensionsGenerator extGen = new ExtensionsGenerator();
				extGen.addExtension(Extension.subjectAlternativeName, false, generalNames);
				requestBuilder.addAttribute(PKCSObjectIdentifiers.pkcs_9_at_extensionRequest, extGen.generate());
			}

			PKCS10CertificationRequest request = requestBuilder.build(contentSigner);

			StringWriter csrWriter = new StringWriter();
			JcaPEMWriter csrPemWriter = new JcaPEMWriter(csrWriter);
			csrPemWriter.writeObject(request);
			csrPemWriter.close();

			return ImmutablePair.of(keyWriter.toString(), csrWriter.toString());

		} catch (Exception e) {
			logger.error(e.getMessage(), e);
			throw new RuntimeException(e);
		}
	}

```

Step 2: Make http call to vault to fetch cert in pem format, using BCJSSE

Step 3: Use the cert and key to make BCFKS file using the code posted in my first thread.


As I mentioned, for now, I will be using keytool command. I am not blocked, you may close the issue.

**Comment by SubhraB** [comment]
@peterdettman 

Just a quick update! 

BCFKS keystores do not work properly as a server side keystore (at least with spring boot), if they have an SAN in it.

Once I removed the SAN from the keystore, everything started working fine. No errors.

**Comment by peterdettman** [comment]
Perhaps you are not adding the SAN extension correctly, or not understanding how it is used. I see from the code example that you are adding dNSName entries. Do you understand that when any dNSName is present, the common name is then ignored and will not be considered for hostname validation purposes?

**Comment by SubhraB** [comment]
I also tried with plain openssl and keytool command generated bcfks, same issue. Without san, works fine, with san, same cipher suite error.

**Comment by peterdettman** [comment]
OK, well if you can provide us a sample non-working keystore, we're happy to look at why it's failing.

**Comment by SubhraB** [comment]
@peterdettman 

I think we are good. I digged further and found the problem. There are two issues. None of them have to do anything with cipher suites.

1) If we use an alias in our keystore which has dot's (.) inside it, the ProvX509KeyManager instance cannot retrieve the private key due to the below if condition in line 274 of ProvX509KeyManager class. I replaced dots with dashed, it started working. So earlier when I used keytool command, the alias was '1', so it was working.

```
                String[] parts = alias.split("\\.");
                if (parts.length == 3)

```

2) Second issue with SAN. This is not related to bouncy castle at all. This is an issue with spring boot (2.2.0), the way they implemented their SSLContextFactory. In They are replacing ProvX509KeyManager instance with their own SniX509ExtendedKeyManager instance. And that too, is not able to retrieve private key since the keystore is not in a regular jks format.

Anyways thanks for your time, I am posting this if it helps someone else in the future.

2)
